### PR TITLE
Multiinstance plus nck fixes

### LINKED ISSUE #309: Some desired hist file comparisons are not done for _N2 tests; also, confusing results in .base.cprnc.out
This issue does NOT appear to be new to the ESMCI version of cime, but I just noticed it.

It looks like some desired hist file comparisons are not done for the .base vs. .rest file comparison for ERS_N2 tests (and I suspect the same is true for other two-run tests).

With cime5.0.3 (in the CESM alpha01 sandbox), I ran ERS_D_N2.f10_f10.ICLM45.yellowstone_gnu.clm-default; I also ran the same test out of clm4_5_8_r182 (with cime4).

In both, it appears that the only history file comparisons that were done (based on the *.rest.cprnc.out files that were produced) were the clm2_0002 hist files. In particular, I see no evidence that comparisons were done for clm2_0001 hist files (a relatively minor issue, but I believe that should be fixed), or for cpl hist files (a more significant issue, I believe).

I was also confused by the presence of *.base.cprnc.out files in these tests, despite the fact that I didn't ask for baseline comparisons. It looks like these compared the clm2_0001 .base history files with the clm2_0002 .base history files. Perhaps this is an intentional feature of _N2 tests. If so, that's fine, but it's confusing for these results to appear in the .base.cprnc.out file, which typically are used for comparisons with baselines. Again, this behavior seems to be the same in cime5 and cime4.

### LINKED ISSUE #310: Multi-instance log files not zipped or copied to $CASEROOT/logs
This issue is NOT new to cime5, but I just noticed it. (I see the same issue with cime5.0.3 and with cime4.) I have observed this with ERS_D_N2.f10_f10.ICLM45.yellowstone_gnu.clm-default.

The multi-instance log files (files like lnd_0001.log.*, lnd_0002.log.*, etc.) are not gzipped, and are not copied to the $CASEROOT/logs directory.

### LINKED ISSUE #443: Convert NCK test to SystemTestsCompareTwo infrastructure
We want to convert the various system tests that compare two runs to use the SystemTestsCompareTwo infrastructure, where possible. I am opening separate issues for each such system test, so that we can check them off as we do this conversion.

### LINKED ISSUE #449: hist_utils.py: _get_latest_hist_files not general enough
I'm continuing my review of PR #413 .... Since that has already been brought to master, I'm opening issues for additional things I notice in this continued review.

hist_utils.py: _get_latest_hist_files (which I believe was written by @jedwards4b ) has this logic:

```python
    date_match = re.compile(r"(\d\d\d\d)-(\d\d)-(\d\d)-(\d\d\d\d\d).nc")
# Some code omitted
            (yr,month,day,time) = s1.group(1,2,3,4)
            s2 = date_match.search(latest_files[ext])
            (pyr,pmonth,pday,ptime) = s2.group(1,2,3,4)
            if yr > pyr or (yr == pyr and month > pmonth) or \
                    (yr == pyr and month == pmonth and day > pday) or \
                    (yr == pyr and month == pmonth and day == pday and time > ptime):
                latest_files[ext] = hist
```

This date_match regular expression isn't general enough to match all history files. For example, monthly history files have dates of the form yyyy-mm.

However, my sense is that this logic for date_match and the sort that's done based on checking year, month, day and time is unnecessary here: As long as _get_all_hist_files returns a sorted list, which is currently the case, then I'm pretty sure that the sort will already have taken care of ordering files correctly by date.

So I'd suggest getting rid of the date_match logic, which would make the core loop in this function simply look like:

```python
    for hist in test_hists:
        ext = get_extension(model, hist)
        latest_files[ext] = hist
```

If I'm missing something, and in fact this date match is important, then it should be generalized to work with monthly history files and possibly others.

### LINKED ISSUE #452: All multi-instance history files should be compared in NCK test
Until PR #413 the shell script-based component_compare_test had logic specific to the multi-instance tests (NCK and its variants) so that it would compare both the _0001 and the _0002 file from the multi-instance run with the standard file from the non-multi-instance run. I believe that was a critical piece of these tests. @mvertens wrote this.

That functionality is now gone.

I'll be the first to admit that the old way of doing it wasn't ideal - i.e., embedding this test-specific logic in the general component_compare_test. Ideally, we'd have a mechanism for an individual test to specify some custom code for doing in-test comparisons. (And, as an aside, that would address a desire that I and others have expressed for a while: we'd like to be able to make component-specific tests that do something more than just doing a blanket comparison of history files for equality.) But I realize that that might require some significant rework (or maybe not... I'm not sure).

In any case, I believe we should have some mechanism for maintaining the old NCK (etc.) behavior, even if it's via a short-term kludge that we will improve later.

### LINKED ISSUE #454: Incorrect error reported in test baseline comparison: file had no counterpart
I ran a test with a baseline comparison:

```
[hobart:~/cesm_code/cesm2_0_alpha01/cime/scripts]$ CIME_MODEL=cesm ./create_test NCK_P24x1_Ld3.f10_f10.ICLM45.hobart_nag.clm-default --compare testtag --baseline-root $scratch/cesm_baselines --test-root .
```

and the baseline comparisons failed with this message:

```
2016-08-24 15:45:31: Comparing hists for case 'NCK_P24x1_Ld3.f10_f10.ICLM45.hobart_nag.clm-default.C.20160824_213601' dir1='/scratch/cluster/sacks/NCK_P24x1_Ld3.f10_f10.ICLM45.hobart_nag.clm-default.C.20160824_213601/run', suffix1='',  dir2='/scratch/cluster/sacks/cesm_baselines/testtag/NCK_P24x1_Ld3.f10_f10.ICLM45.hobart_nag.clm-default' suffix2=''
  comparing model 'datm'
    no hist files found for model datm
  comparing model 'clm'
    File '/scratch/cluster/sacks/NCK_P24x1_Ld3.f10_f10.ICLM45.hobart_nag.clm-default.C.20160824_213601/run/NCK_P24x1_Ld3.f10_f10.ICLM45.hobart_nag.clm-default.C.20160824_213601.clm2.h0.0001-01-04-00000.nc' had no counterpart in '/scratch/cluster/sacks/cesm_baselines/testtag/NCK_P24x1_Ld3.f10_f10.ICLM45.hobart_nag.clm-default' with suffix ''
    File '/scratch/cluster/sacks/NCK_P24x1_Ld3.f10_f10.ICLM45.hobart_nag.clm-default.C.20160824_213601/run/NCK_P24x1_Ld3.f10_f10.ICLM45.hobart_nag.clm-default.C.20160824_213601.clm2.h1.0001-01-04-00000.nc' had no counterpart in '/scratch/cluster/sacks/cesm_baselines/testtag/NCK_P24x1_Ld3.f10_f10.ICLM45.hobart_nag.clm-default' with suffix ''
```

This is despite the fact that the baseline directory contains the appropriate files:

```
[hobart:/scratch/cluster/sacks/cesm_baselines/testtag/NCK_P24x1_Ld3.f10_f10.ICLM45.hobart_nag.clm-default]$ ls -l
total 72012
drwxrwxr-x 2 sacks cgdtss     4096 Aug 24 14:12 CaseDocs
-rw------- 1 sacks cgdtss  1290812 Aug 24 14:24 clm2.h0.0001-01-04-00000.nc
-rw------- 1 sacks cgdtss  4810332 Aug 24 14:24 clm2.h1.0001-01-04-00000.nc
-rw------- 1 sacks cgdtss 67602892 Aug 24 14:24 cpl.hi.0001-01-04-00000.nc
-rw-r--r-- 1 sacks cgdtss     5863 Aug 24 14:24 cpl.log.gz
-rw-rw---- 1 sacks cgdtss     1218 Aug 24 14:12 user_nl_clm
-rw-rw---- 1 sacks cgdtss     2102 Aug 24 14:12 user_nl_cpl
-rw-rw---- 1 sacks cgdtss      787 Aug 24 14:12 user_nl_datm
-rw-rw---- 1 sacks cgdtss      573 Aug 24 14:12 user_nl_rtm
```

I can't figure out why the comparisons aren't finding the appropriate baseline hist file. It's possible that I did something wrong here, though I can't see anything amiss. At first I thought this was specific to the multi-instance test, although it doesn't seem like it's trying to compare multi-instance (_0001 or _0002) files as far as I can tell.

### LINKED ISSUE #468: Cannot create multiple tests that write baselines to the same directory
It appears that the new logic to prevent you from overwriting existing baselines is too aggressive.

The way it's supposed to work is: If you have a baseline for (e.g.) test SMS.f45_g37_rx1.A.yellowstone_intel in baseline directory cesm2_0_alpha01e, then you shouldn't be able to re-generate baselines for that same test in the same baseline directory.

However, the way it's currently operating is that it prevents you from writing ANY new baselines (even for other tests) to an existing baseline directory.

Steps to reproduce:

```
./create_test --no-build SMS.f45_g37_rx1.A.yellowstone_intel --generate foo --baseline-root /glade/scratch/$USER/cesm_baselines

./create_test --no-build SMS.f45_g37_rx1.A.yellowstone_pgi --generate foo --baseline-root /glade/scratch/$USER/cesm_baselines
```

The first command succeeds (as expected), but the second fails with:

```
ERROR: Refusing to overwrite existing baseline directory, use -o to force overwrite /glade/scratch/sacks/cesm_baselines/foo
```

This error should only have been given if I tried to rerun the exact same test - i.e., the intel test.

### LINKED ISSUE #472: This line of code looks wrong in hist_utils.py: _hists_match
In this code snippet:

```python
    for hists, suffix, normalized in [(hists1, suffix1, normalized1), (hists2, suffix2, normalized2)]:
        for hist in hists:
            normalized_name = hist[hist.rfind(model):]
            if suffix1 != "":
                expect(normalized_name.endswith(suffix), "How did '%s' not have suffix '%s'" % (hist, suffix))
                normalized_name = normalized_name[:len(normalized_name) - len(suffix)]
```

The use of suffix1 in the conditional:

```python
            if suffix1 != "":
```

looks wrong to me: Intuitively, it looks like this should be simply `suffix`. I haven't taken the time to fully understand this function, so my intuition may be wrong... but if it is, an explanatory comment would help here.

### PR DESCRIPTION
**Pull request by jedwards4b**
Combines PR #471 and #465 

Test suite: scripts_regression_tests, NCK_D_Ld3.f10_f10.ICLM45, ERS_D_N2_Ld3.f10_f10.ICLM45
Test baseline: 
Test namelist changes: 
Test status: bit for bit

Fixes #443 
Fixes #309
Fixes #310 
Fixes #449 
Fixes #452 
Fixes #454
Fixes #468 
Fixes #472 
User interface changes?: 

Code review:

### CODE REVIEW

**Comment by jedwards4b** [comment]
Testing in progress, do not merge yet - but please review.

File: utils/python/CIME/hist_utils.py [hunk 1]
```diff
@@ -123,9 +126,40 @@ def _hists_match(model, hists1, hists2, suffix1="", suffix2=""):
...
+                for idx1, hist2 in enumerate(hists2):
+                    norm_hist2 = normalized2[idx1]
+                    if norm_hist1 == norm_hist2:
+                        match_ups.append((hists1[idx], hist2))
+                        if hist2 in two_not_one:
+                            two_not_one.remove(hist2)
+                        if hists1[idx] in one_not_two:
+                            one_not_two.remove(hists1[idx])
+        # in this case hists2 contains multiinstance hists1 does not
+        print multi_normalized2
```
Comment (jgfouca) [inline utils/python/CIME/hist_utils.py]: I assume these prints were for debugging. They can be removed now.

**Comment by jgfouca** [comment]
I only see one minor problem. Everything else looks good.

**Comment by jedwards4b** [comment]
I've rebased to master to fix the merge issue and tested, there are pylint issues in scripts_regression_tests which are present on the current trunk.

**Comment by jedwards4b** [comment]
All tests now pass.

**Comment by jgfouca** [comment]
OK, ready to merge? @billsacks ?

**Comment by jedwards4b** [comment]
Sorry - he gave me the okay via email.   Already merged.

**Comment by jgfouca** [comment]
Ah, didn't notice. No problem

**Pull request merged**